test/tests/typeguards.test.ts

```diff
@@ -0,0 +1,43 @@
+import { expect } from 'chai';
+import { isDocument, isDocumentArray } from '../../src/utils';
+import { UserRefModel } from '../models/userRefs';
+
+export function typeguards() {
+    it ('should guarantee array of document types', async () => {
+        const UserMaster = await UserRefModel.create({
+            name: 'master'
+        });
+        const UserSub = await UserRefModel.create({
+            name: 'sub',
+            master: UserMaster._id
```

Comment:
```suggestion
            master: UserMaster
```
It works without `._id` and is more clean

Reply:
in practice this almost never happens, you have the ref and you want to explicitly populate, it's this way in order to work as example too

Reply:
to improve the test as example i will add the _id in another line

Reply:
in everything i so far used typegoose in, it worked so, so i want to ask now, in which cases does it *not* work?

Reply:
Then you are simply not using populate and with type safety accessing it's properties, you are using the refs as parameters for findById and filling the documents manually, is another way of doing it, but much more verbose

Reply:
when making it without id, mongoose just greps the model's, so it is the same... (i tested it always with type safety, thats why i asked "in which case this does *not* work"(this = without `._id`)

Reply:
and as an example: 
look at [#204](https://github.com/szokodiakos/typegoose/issues/204#issuecomment-515482936)

with the addition:
```ts
const found = await CPModel.findById(cp.id).populate("car").populate("buyer").exec();
// console.log(found);

console.log("buyer", isDocument(found.buyer));
console.log("car", isDocument(found.car));
```
with the output of
```
buyer true
car true
```

and in the DB
```json
{
  "_id": "5d3b137acceba85ceffac79f",
  "buyer": "5d3b137acceba85ceffac79d",
  "car": "5d3b137acceba85ceffac79c",
  "referral": {
    "referralDiscount": 1,
    "referrer": "5d3b137acceba85ceffac79d"
  },
  "__v": 0
}
```

 so i dont really understand where it should *not* work

Reply:
You are not accessing the car and buyer properties, which is the whole purpose of the type guards, and giving the whole document as parameter or field is just not how refs really work, in population mongoose looks for the _ID

Reply:
Try to do
```typescript
console.log(found.buyer.name);
console.log(found.car.model);
```

Reply:
> when making it without id, mongoose just greps the model's, so it is the same... (i tested it always with type safety, thats why i asked "in which case this does _not_ work"(this = without `._id`)

the fact that mongoose handles it automatically does not mean it is doing the `.populate` function, and no matter how are you working around it, you can't access the properties of the populated document field without a type guard, which is in fact a documented feature of typescript
 [union type](https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types)
 [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types)

> Ref\<T\>
> ```javascript
> export type Ref<T> = T | mongoose.Schema.Types.ObjectId;
> ```

Reply:
I think we are talking past each other, i mean that `{ buyer: user, car, referral: rfl }` is the same as `{ buyer: user.id, car: car.id, referral: rfl.id }`, because they give the same result in the DB, and thus not modifying how the populate and with that the type guard works, my question was just about the `.id` and not about the type guard, because the type guard would work so or so

Reply:
ok, i understand, but to explain better the use case in the test i just prefer to use populate explicitly, that's it

Reply:
*resolved*
thanks for taking the time to discuss about it